packages/sit-onyx/src/components/OnyxDataGrid/features/index.ts

```diff
@@ -0,0 +1,178 @@
+import type { DataGridEntry, DataGridMetadata } from "../types";
+
+export type TableFeature<TEntry extends DataGridEntry, TFeatureName extends symbol> = {
+  /**
+   * Unique name and identifier of the table feature
+   */
+  name: TFeatureName;
+
+  /**
+   * An array of reactive states that should trigger a table re-generation
+   */
+  watch: WatchSource[];
+
+  /**
+   * Allows modifying the table state as a whole.
+   */
+  mutation?: {
+    func: (state: TEntry[]) => void;
+  };
+
+  /**
+   * Allows the modification of the headers.
+   */
+  header?: {
+    /**
+     * Adds header icon button(s).
+     * `iconComponent` of an action is shown after the header label.
+     * The components must be ARIA-conform buttons.
+     */
+    actions?: (column: keyof TEntry) => {
+      iconComponent: Component;
+      listComponent?: never;
+    }[];
+  };
+};
+
+/**
+ * Helper function that infers the generics of the TableFeature type.
+ * @example
+ * ```ts
+ *
+ * const MY_FEATURE = Symbol("TABLE_HEADER_BUTTON");
+ * export const useTableHeaderButton = createFeature(<TEntry extends DataGridEntry>() => {
+ *   return {
+ *     name: MY_FEATURE,
+ *     header: {
+ *       actions: (column) => [
+ *         {
+ *           iconComponent: h('button', { onClick: (column) => console.log(`Clicked on ${column}`) },),
+ *         },
+ *       ],
+ *     },
+ *   };
+ * });
+ * ```
+ */
+export const createFeature = <TFeatureName extends symbol, TArgs extends unknown[]>(
+  featureDefinition: <TEntry extends DataGridEntry>(
+    ...args: TArgs
+  ) => TableFeature<TEntry, TFeatureName>,
+) => featureDefinition;
+
+type ExtractTEntry<T> = T extends TableFeature<infer I, symbol>[] ? I : never;
+
+/**
+ * Uses the defined table features to provide factory functions.
+ * These factories are to be used to map data and configuration to `OnyxDataGridRenderer` properties.
+ * The properties are then used to render the data grid.
+ *
+ * Make use of the `watchSources` to trigger re-rendering when state changes occur.
+ * @example
+ * ```vue
+ * <script setup lang="ts">
+ * // ...
+ * // imports, props, emits, etc.
+ * const withHeaderButton = useTableHeaderButton<TEntry>();
+ *
+ * const { watchSources, createRendererRows, createRendererColumns } = useTableFeatures([withSorting]);
+ *
+ * const renderCols: Ref<DataGridRendererColumn<TEntry, object>[]> = ref([]);
+ * const renderRows: Ref<DataGridRendererRow<TEntry, DataGridMetadata>[]> = ref([]);
+ *
+ * const { columns, data } = toRefs(props);
+ *
+ * watch(
+ *   [columns, data, ...watchSources],
+ *   ([newColumns, newData]) => {
+ *     renderCols.value = createRendererColumns(newColumns);
+ *     renderRows.value = createRendererRows(newData, newColumns);
+ *   },
+ *   { immediate: true },
+ * );
+ * </script>
+ * <template>
+ *   <OnyxDataGridRenderer :columns="renderCols" :rows="renderRows" />
+ * </template>
+ * ```
+ */
+export const useTableFeatures = <
+  T extends TableFeature<TEntry, symbol>[] | [],
```

Comment:
```suggestion
  T extends TableFeature<TEntry, symbol>[] | [],
```

Why is the `| []` needed?